include/simdjson/generic/ondemand/document.h

```diff
@@ -178,24 +180,38 @@ class document {
    * @returns A value of the given type, parsed from the JSON.
    * @returns INCORRECT_TYPE If the JSON value is not the given type.
    */
-  template<typename T> simdjson_inline simdjson_result<T> get() & noexcept {
+  template<typename T> simdjson_inline simdjson_result<T> get() &
+#ifdef __cpp_concepts
+    noexcept(tag_invocable<deserialize_tag, std::type_identity<T>, document&> ? nothrow_tag_invocable<deserialize_tag, std::type_identity<T>, document&> : true)
+#else
+    noexcept
+#endif
+  {
+#ifdef __cpp_concepts
+    if constexpr (tag_invocable<deserialize_tag, std::type_identity<T>, document&>) {
+      return deserialize(std::type_identity<T>{}, *this);
+    } else {
+#endif // __cpp_concepts
     // Unless the simdjson library or the user provides an inline implementation, calling this method should
     // immediately fail.
     static_assert(!sizeof(T), "The get method with given type is not implemented by the simdjson library. "
       "The supported types are ondemand::object, ondemand::array, raw_json_string, std::string_view, uint64_t, "
       "int64_t, double, and bool. We recommend you use get_double(), get_bool(), get_uint64(), get_int64(), "
       " get_object(), get_array(), get_raw_json_string(), or get_string() instead of the get template."
       " You may also add support for custom types, see our documentation.");
+#ifdef __cpp_concepts
+      }
+#endif
   }
   /** @overload template<typename T> simdjson_result<T> get() & noexcept */
-  template<typename T> simdjson_inline simdjson_result<T> get() && noexcept {
-    // Unless the simdjson library or the user provides an inline implementation, calling this method should
-    // immediately fail.
-    static_assert(!sizeof(T), "The get method with given type is not implemented by the simdjson library. "
-      "The supported types are ondemand::object, ondemand::array, raw_json_string, std::string_view, uint64_t, "
-      "int64_t, double, and bool. We recommend you use get_double(), get_bool(), get_uint64(), get_int64(), "
-      " get_object(), get_array(), get_raw_json_string(), or get_string() instead of the get template."
-      " You may also add support for custom types, see our documentation.");
+  template<typename T> simdjson_inline simdjson_result<T> get() &&
+#ifdef __cpp_concepts
+    noexcept(tag_invocable<deserialize_tag, std::type_identity<T>, document&> ? nothrow_tag_invocable<deserialize_tag, std::type_identity<T>, document&> : true)
+#else
+    noexcept
+#endif
+  {
+      return static_cast<document&>(*this).get<T>();
```

Comment:
That's potentially a problem, see https://github.com/simdjson/simdjson/pull/2229

It might allow you to hold an `ondemand::array` or `ondemand::object`  while the `ondemand::document` is out of scope.

The workaround might be to check whether `T` is either  `ondemand::array` or `ondemand::object` and error in these cases. That could be achieved with a `static_assert`?

What do you think?

Reply:
Honestly I freaked out when I saw this:

https://github.com/simdjson/simdjson/blob/a33fdc2770549ec1f602736420c2e1b18b7df2f0/include/simdjson/generic/ondemand/document-inl.h#L170-L176

1. we're returning string_views from a moved type (which I assume it points to the parser not document itself)
2. I don't understand why `std::forward` is being used, and I also why all of them don't have it

If the problem is only array and object, then `static_assert` would be better than the code that I mentioned that we're simply not providing specializations for array and object which the user may add themselves later!

Reply:
> we're returning string_views from a moved type (which I assume it points to the parser not document itself)

That is fine because the `std::string_view` is owned by the parser.

> I don't understand why std::forward is being used, and I also why all of them don't have it

I suspect that the `std::forward<document>(*this).get_int64();` is overcomplicated code. It could be simplified.